### PR TITLE
Issue 58 - use requestIdleCallback when available for whenRouteIdle

### PR DESCRIPTION
**Pull request by scalvert**
First pass at implementing #58.

TODO:
- [ ]  figure out if there's a way to optimize all the wrapping functions

### CODE REVIEW

File: tests/unit/capabilities-test.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { module, test } from 'qunit';
+import Capabilities from 'ember-app-scheduler/capabilities';
+
+const CAPABILITIES = Capabilities.instance;
+
+module('Unit | Capabilities', function(hooks) {
+  hooks.beforeEach(function() {
+    debugger;
```
Comment (rwjblue) [inline tests/unit/capabilities-test.js]: 🔪

File: addon/capabilities.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  constructor() {
+    if (_instance) {
+      throw new Error('Capabilities can only be constructed once');
+    }
+
+    this.reset();
+    this.DISABLED = true;
+  }
+
+  static get instance() {
```
Comment (rwjblue) [inline addon/capabilities.js]: Maybe we can kill this (and the instance check in constructor), and instead just have our default export be the instance?

```js
export class Capabilities { /* stuff here */ }

export default new Capabilities();
```

File: addon/capabilities.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+let _instance = null;
+
+class Capabilities {
+  constructor() {
+    if (_instance) {
+      throw new Error('Capabilities can only be constructed once');
+    }
+
+    this.reset();
+    this.DISABLED = true;
```
Comment (rwjblue) [inline addon/capabilities.js]: This is confusing, took me a few read throughs of the diff to grok what was even happening.

Do we really need it? Can't we call `.overrideCapabilities('requestIdleCallback', true)`?

Comment (scalvert) [reply addon/capabilities.js]: Of course we can. I added it for readability, but I think it's bad to have it as part of the class. I could export another var. Personally, I dislike booleans passed into methods as it doesn't do a good job informing people as to what the parameter controls.

File: addon/capabilities.js (hunk 2 above)
Comment (scalvert) [reply addon/capabilities.js]: So I did that, but was seeing issues in tests where I required the module, and would get a new instance of Capabilities.

File: tests/unit/capabilities-test.js (hunk 1 above)
Comment (scalvert) [reply tests/unit/capabilities-test.js]: Good catch.

File: addon/capabilities.js (hunk 2 above)
Comment (rwjblue) [reply addon/capabilities.js]: the module should only be evaluated once, so `export default new Thing()` should only ever create a single instance

happy to dig in with you to see whats going on

**Pull request merged**